In [3]:
import pandas as pd
import os, sys


file_dir = os.path.dirname(os.path.abspath("__file__"))
parent_dir = os.path.dirname(file_dir)
newPath = os.path.join(parent_dir, 'data/db_historic/rt_trips_2016_I_DB.txt')

# read leavetimes 2016 .txt file into csv leave16 (just first 100)
trips16 = pd.read_csv(newPath, delimiter=';')


trips16

datasource        dayofservice   tripid lineid  routeid  direction  \
0              DB  09-FEB-16 00:00:00  2824642    145  145_105          2   
1              DB  09-FEB-16 00:00:00  2826717      9      9_7          2   
2              DB  09-FEB-16 00:00:00  2826730    54A   54A_12          2   
3              DB  09-FEB-16 00:00:00  2826743      7     7_51          1   
4              DB  09-FEB-16 00:00:00  2812908     39    39_20          1   
5              DB  09-FEB-16 00:00:00  2813717    56A   56A_30          2   
6              DB  09-FEB-16 00:00:00  2815965     37    37_14          1   
7              DB  09-FEB-16 00:00:00  2815967     11    11_40          1   
8              DB  09-FEB-16 00:00:00  2815970     11    11_42          2   
9              DB  09-FEB-16 00:00:00  2818038     63    63_18          1   
10             DB  09-FEB-16 00:00:00  2818052    122   122_16          2   
11             DB  09-FEB-16 00:00:00  2818069    39A   39A_41          1   
12             DB  09-FEB-16 00:00:00  2818078     70    70_60          1   
13             DB  09-FEB-16 00:00:00  2818165     42    42_42          1   
14             DB  09-FEB-16 00:00:00  2818169     42    42_42          1   
15             DB  09-FEB-16 00:00:00  2818172     42    42_44          2   
16             DB  09-FEB-16 00:00:00  2811524    40D  40D_104          2   
17             DB  09-FEB-16 00:00:00  2811530     32    32_57          1   
18             DB  09-FEB-16 00:00:00  2811532    31A   31A_26          2   
19             DB  09-FEB-16 00:00:00  2811561     40    40_17          1   
20             DB  09-FEB-16 00:00:00  2811563     40    40_17          1   
21             DB  09-FEB-16 00:00:00  2811564     40    40_24          2   
22             DB  09-FEB-16 00:00:00  2811618     67     67_6          2   
23             DB  09-FEB-16 00:00:00  2811645    40B   40B_65          2   
24             DB  09-FEB-16 00:00:00  2819337    41C   41C_79          2   
25             DB  09-FEB-16 00:00:00  2812750    38A   38A_32          2   
26             DB  09-FEB-16 00:00:00  2812756    123   123_34          1   
27             DB  09-FEB-16 00:00:00  2812758    123   123_34          1   
28             DB  09-FEB-16 00:00:00  2821361    123   123_36          2   
29             DB  09-FEB-16 00:00:00  2821464     40    40_17          1   
...           ...                 ...      ...    ...      ...        ...   
929477         DB  24-MAY-16 00:00:00  3044682    151   151_17          2   
929478         DB  24-MAY-16 00:00:00  3037424     11    11_42          2   
929479         DB  24-MAY-16 00:00:00  3037448      7     7_51          1   
929480         DB  24-MAY-16 00:00:00  3037452      7     7_60          1   
929481         DB  24-MAY-16 00:00:00  3047012     42    42_42          1   
929482         DB  24-MAY-16 00:00:00  3047017     14    14_15          1   
929483         DB  24-MAY-16 00:00:00  3047061    41C   41C_79          2   
929484         DB  24-MAY-16 00:00:00  3039581     44    44_36          2   
929485         DB  24-MAY-16 00:00:00  3039613    185   185_61          2   
929486         DB  24-MAY-16 00:00:00  3053304     49    49_27          1   
929487         DB  24-MAY-16 00:00:00  3036348     79    79_11          2   
929488         DB  24-MAY-16 00:00:00  3036390     49    49_27          1   
929489         DB  24-MAY-16 00:00:00  3036471     37    37_14          1   
929490         DB  24-MAY-16 00:00:00  3036474     37    37_15          2   
929491         DB  24-MAY-16 00:00:00  3048415    130   130_11          2   
929492         DB  24-MAY-16 00:00:00  3048430    39A   39A_43          2   
929493         DB  24-MAY-16 00:00:00  3049413     15    15_16          1   
929494         DB  24-MAY-16 00:00:00  3044171     25   25_269          2   
929495         DB  24-MAY-16 00:00:00  3044183     25   25_269          2   
929496         DB  24-MAY-16 00:00:00  3044246    150    150_6          2   
929

In [4]:
# See how many rows and columns are in it
trips16.shape

(929507, 16)

In [5]:
trips16.dtypes

datasource          object
dayofservice        object
tripid               int64
lineid              object
routeid             object
direction            int64
plannedtime_arr      int64
plannedtime_dep      int64
actualtime_arr     float64
actualtime_dep     float64
basin               object
tenderlot          float64
suppressed         float64
justificationid    float64
lastupdate          object
note                object
dtype: object

In [6]:
#Missing data all features
trips16.isnull().sum()

datasource              0
dayofservice            0
tripid                  0
lineid                  0
routeid                 0
direction               0
plannedtime_arr         0
plannedtime_dep         0
actualtime_arr      75527
actualtime_dep      87804
basin                   0
tenderlot          929507
suppressed         927797
justificationid    927798
lastupdate              0
note                    0
dtype: int64

In [7]:
trips16.describe().T

count          mean            std        min        25%  \
tripid           929507.0  3.004382e+06  141708.913412  2777084.0  2888369.0   
direction        929507.0  1.504126e+00       0.499983        1.0        1.0   
plannedtime_arr  929507.0  5.527825e+04   17354.882735    18539.0    39931.0   
plannedtime_dep  929507.0  5.163129e+04   17521.013791    16200.0    36120.0   
actualtime_arr   853980.0  5.500930e+04   17283.672078    17576.0    39663.0   
actualtime_dep   841703.0  5.183067e+04   17509.413476    16090.0    36603.0   
tenderlot             0.0           NaN            NaN        NaN        NaN   
suppressed         1710.0  8.239766e-01       0.380951        0.0        1.0   
justificationid    1709.0  2.205045e+04    4981.588208      555.0    19140.0   

                       50%        75%        max  
tripid           3011742.0  3039788.0  3293472.0  
direction              2.0        2.0        2.0  
plannedtime_arr    55560.0    69179.0    91920.0  
plannedtime_dep    51600.0    65400.0    88200.0  
actualtime_arr     55208.0    68918.0    92142.0  
actualtime_dep     51712.0    65673.0    97198.0  
tenderlot              NaN        NaN        NaN  
suppressed             1.0        1.0        1.0  
justificationid    19240.0    21538.0    31632.0

In [8]:
# check for duplicate rows
print('Duplicate rows:', trips16.duplicated()[trips16.duplicated() == True].shape[0])
# Check for duplicate columns
print('Duplicate columns:',trips16.columns.size - trips16.columns.unique().size)

Duplicate rows: 0
Duplicate columns: 0


In [9]:
# Check for irregular cardinality. 
print("Unique values for:\n- datasource:", pd.unique(trips16.datasource.ravel()))
#print("\n- dayofservice:", sorted(pd.unique(trips16.dayofservice.ravel())))
print()
#print("\n- tripid:", sorted(pd.unique(trips16.tripid.ravel())))
print()
print("\n- lineid:", sorted(pd.unique(trips16.lineid.ravel())))
print()
print("\n- routeid:", sorted(pd.unique(trips16.routeid.ravel())))
print()
print("\n- direction:", sorted(pd.unique(trips16.direction.ravel())))
print()
#print("\n- plannedtime_arr:", sorted(pd.unique(trips16.plannedtime_arr.ravel())))
print()
#print("\n- plannedtime_dep:", sorted(pd.unique(trips16.plannedtime_dep.ravel())))
print()
#print("\n- actualtime_arr:", sorted(pd.unique(trips16.actualtime_arr.ravel())))
print()
print()
#print("\n- actualtime_dep:", sorted(pd.unique(trips16.actualtime_dep.ravel())))
print("\n- basin:", sorted(pd.unique(trips16.basin.ravel())))
print()
print("\n- tenderlot:", sorted(pd.unique(trips16.tenderlot.ravel())))
print()
print("\n- suppressed:", sorted(pd.unique(trips16.suppressed.ravel())))
print()
print("\n- justificationid:", sorted(pd.unique(trips16.justificationid.ravel())))
print()
#print("\n- lastupdate:", sorted(pd.unique(trips16.lastupdate.ravel())))



Unique values for:
- datasource: ['DB']



- lineid: ['1', '102', '104', '11', '111', '114', '116', '118', '120', '122', '123', '13', '130', '14', '140', '142', '145', '14C', '15', '150', '151', '15A', '15B', '16', '161', '16C', '17', '17A', '18', '184', '185', '220', '236', '238', '239', '25', '25A', '25B', '25D', '25X', '26', '27', '270', '27A', '27B', '27X', '29A', '31', '31A', '31B', '32', '32X', '33', '33A', '33B', '33X', '37', '38', '38A', '38B', '38D', '39', '39A', '4', '40', '40B', '40D', '41', '41A', '41B', '41C', '41X', '42', '43', '44', '44B', '45A', '46A', '46E', '47', '49', '51D', '51X', '53', '54A', '56A', '59', '61', '63', '65', '65B', '66', '66A', '66B', '66X', '67', '67X', '68', '68A', '69', '69X', '7', '70', '75', '757', '76', '76A', '77A', '77X', '79', '79A', '7B', '7D', '8', '83', '83A', '84', '84A', '84X', '9']


- routeid: ['102_10', '102_8', '102_9', '104_15', '104_16', '111_3', '111_4', '114_5', '114_6', '116_1', '116_3', '118_3', '11_40', '11_41', '11_42', '11_

In [10]:
# Return the cardinality of the set of unique values, i.e., number of distinct values for feature x
len(trips16['dayofservice'].unique())


153

In [11]:
len(trips16['lineid'].unique())

120

In [12]:
len(trips16['tripid'].unique())

194348

In [13]:
len(trips16['routeid'].unique())

481

In [14]:
len(trips16['plannedtime_arr'].unique()) 

51685

In [15]:
len(trips16['plannedtime_dep'].unique()) 

768

In [16]:
len(trips16['actualtime_arr'].unique()) 


66332

In [17]:
len(trips16['actualtime_dep'].unique()) 

64626

In [18]:
len(trips16['tenderlot'].unique()) 

1

In [19]:
len(trips16['suppressed'].unique()) 

3

In [20]:
len(trips16['justificationid'].unique())

174

In [21]:
len(trips16['lastupdate'].unique())


153

'Suppressed' is the feature that might be most relevant in data cleaning. Of the 929507 rows in the entire dataframe, 927797 are NaN for suppressed. With just 1710 having actual values (1.0 or 0.0). The frequency tables below show that there are 1409 rows with 1.0 and 301 with 0.0. (frequency tables don't seem to record NaN values).

Here is what is said in the NTA pdf about the feature 'suppressed':
"If the trip is planned and achieved, the record will contain both the planned and actual data. If the
tripped is planned and suppressed the record will contain only the planned data and the flag
“Suppressed” will be checked.
If the trips is partially suppressed the flag “Suppressed” will not be checked. The set of the leave
times suppressed is detailed in table RT_LeaveTimes"

It also says, in explaining suppressed: "The whole trip has been suppressed (0 =achieved, 1 = suppressed)"

So rows with the value 1 for suppressed must be dropped - if they only have data for planned times, not actual times, then they will be of no use in training the model. So this will be dropped when the trips2016 csv is created.




In [22]:
# Get frequency of different levels for value 'suppressed'
frequency = pd.crosstab(index=trips16['suppressed'],  
                              columns="count")      
frequency




col_0       count
suppressed       
0.0           301
1.0          1409

In [23]:
# Get proportion of levels for value 'suppressed'
frequency2 = pd.crosstab(index=trips16['suppressed'],  
                              columns="proportion %")
frequency2/frequency2.sum()*100

col_0       proportion %
suppressed              
0.0            17.602339
1.0            82.397661

In [24]:
# Get frequency of different levels for value 'justificationid'
#frequency = pd.crosstab(index=trips16['justificationid'],  
                              #columns="count")      
#frequency


CONCLUSIONS:
- There are NO duplicate rows or columns in whole of trips 2016 txt file
- datasource column can be dropped as it just has a single value, DB.
- Basin can be dropped (it just has a single value, BasDef, which is of no relevance to our model).
- For the Tenderlot feature, all entries are NaN. So we will drop it.
- although there are some entries in 'justificationid', as the feature records 'fault codes' to do with why a journey was suppressed - this is not relevant to developing our model, so we will be dropping this column. 
- 'note' is not relevant to our model either, so will be dropped 
- 'lastupdate' is not relevant to our model either, as it does not relate to what day a journey took place, only when the data was updated. So we will drop this column completely also.
- We will also be dropping all rows in the suppressed column where suppressed=1
